### ISSUE TITLE
Variant Return nil If user is nil when Flag has more than one variant 

### ISSUE THREAD

**Issue by areejSad** [amplitude/experiment-ios-client#45]
I faced an issue, that the whole variant from  experiment.variant(key.rawValue) return nil.
This happens when I don't want to give a User since it's optional.

![Uploading Screenshot 2024-02-09 at 4.12.36 PM.png…]()
<img width="1217" alt="Screenshot 2024-02-09 at 4 14 44 PM" src="https://github.com/amplitude/experiment-ios-client/assets/152368592/6258a92e-bbb4-4331-847d-40e8c44d49d5">

**Comment by bgiori** [comment]
Hi @areejSad Thanks for reporting this issue.

What version of the SDK are you using? And can you link the flag/experiment in amplitude?

Thanks!

**Comment by areejSad** [comment]
the version '1.13.0', sorry what do you exactly need? if you mean the space it's a private account for company

**Comment by bgiori** [comment]
Is the deployment used to initialize the SDK added to the flag, and is the flag activated?

A link to the flag/experiment in Amplitude would be sufficient for me to check.

**Comment by areejSad** [comment]
I can show you a photo, and yes it's added to the deployment environment, I also checked amplitudeExperimentAPIKey
amplitudeAnalyticsAPIKey 
and it's the same added in the project.
This issue only happened if I have multiple variant, but If I just changed it to one variant it will work fine in the same code. So I'm confused why.

this is a photo to check the flag and it's active 
<img width="1728" alt="Screenshot 2024-02-14 at 12 29 48 AM" src="https://github.com/amplitude/experiment-ios-client/assets/152368592/78ac71d5-d0b2-4a83-bb9a-a0c700b28f26">

**Comment by bgiori** [comment]
Thanks, and sorry for the delay.

>This issue only happened if I have multiple variant, but If I just changed it to one variant it will work fine in the same code. So I'm confused why.

This most likely means that the user information (user_id and device_id) are not being sent in the fetch request. 

Can you turn on the `debug` config option and look for the logs with `[Experiment]` in them? Specifically the on that has `Fetch variants for user`

**Comment by areejSad** [comment]
yes exactly I'm intentionally making (user_id and device_id) as nil but it suppose to be ok right? because it's optional to have a user? and if there's no user it suppose to return a random variant right? 

and this is how I initialize 

<img width="500" alt="Screenshot 2024-02-22 at 1 56 15 PM" src="https://github.com/amplitude/experiment-ios-client/assets/152368592/c9bd058f-27f0-4ab7-9c0f-ab7f89e5ec48">


===========


and this is how I fetch:

<img width="669" alt="Screenshot 2024-02-22 at 1 58 00 PM" src="https://github.com/amplitude/experiment-ios-client/assets/152368592/b7a603bd-580a-44d0-a74b-46c97c14c9fd">

**Comment by areejSad** [comment]
it says in the documentation: 

<img width="895" alt="Screenshot 2024-02-22 at 2 00 49 PM" src="https://github.com/amplitude/experiment-ios-client/assets/152368592/690ea389-16de-400b-9d78-d35f86c99a6e">

**Comment by bgiori** [comment]
> yes exactly I'm intentionally making (user_id and device_id) as nil but it suppose to be ok right? because it's optional to have a user? and if there's no user it suppose to return a random variant right?

Not quite. When there is some form of randomization (e.g. 3 variants evenly distributed) then we need some input value to get a consistent output. Without a user ID or device ID we dont have an input for the hash function used to determine which variant the user should be assigned, so the result is empty.

Passing a nil user is generally if you are using the `initializeWithAmplitudeAnalytics` initialization function or you have intialized the SDK with a custom `UserProvider`
